packages/frontend/src/components/accounts/recovery_setup/SetupRecoveryMethod.js

```diff
@@ -246,7 +245,35 @@ class SetupRecoveryMethod extends Component {
                     });
                 } catch (e) {
                     console.warn(e.code);
-                    await fundCreateAccount(accountId, recoveryKeyPair, method.kind);
+                    const isNewAccount = await checkIsNew(accountId);
+                    if (!isNewAccount) {
+                        const accessKeys = await wallet.getAccessKeys(accountId) || [];
+                        const publicKeys = accessKeys.map(key => key.public_key);
+                        const publicKeyExistsOnAccount = publicKeys.includes(recoveryKeyPair.publicKey.toString());
+
+                        if (publicKeyExistsOnAccount) {
+                            try {
+                                await wallet.saveAndMakeAccountActive(accountId);
+                                await wallet.addLocalKeyAndFinishSetup(accountId, method.kind, recoveryKeyPair.publicKey);
+                            } catch (e) {
+                                showCustomAlert({
+                                    success: false,
+                                    messageCodeHeader: 'error',
+                                    errorMessage: e.message
+                                });
+                                redirectTo('/recover-account');
+                            }
+                        } else {
+                            showCustomAlert({
+                                success: false,
+                                messageCodeHeader: 'error',
+                                messageCode: 'walletErrorCodes.createNewAccount.accountExists.error',
+                                errorMessage: e.message
+                            });
+                        }
+                    } else {
+                        await fundCreateAccount(accountId, recoveryKeyPair, method.kind);
+                    }
                 }
```

Comment:
The level of nesting here is getting a bit out of hand :D. By unrolling some of these if/else and using short-circuits we can make this logic a lot more linear -- e.g. something like this:
```suggestion
                    const isNewAccount = await checkIsNew(accountId);

                    if (isNewAccount) {
                        // No on-chain account exists; fallback to implicit or verified account flow
                        await fundCreateAccount(accountId, recoveryKeyPair, method.kind);
                        return;
                    }

                    // on-chain account exists; verify that it is owned by the current key
                    const accessKeys = await wallet.getAccessKeys(accountId) || [];
                    const publicKeys = accessKeys.map(key => key.public_key);
                    const publicKeyExistsOnAccount = publicKeys.includes(recoveryKeyPair.publicKey.toString());

                    if (!publicKeyExistsOnAccount) {
                        // Someone else must've created the account (or a two-tabs-open situation)
                        showCustomAlert({
                            success: false,
                            messageCodeHeader: 'error',
                            messageCode: 'walletErrorCodes.createNewAccount.accountExists.error',
                            errorMessage: e.message
                        });
                        return;
                    }

                    // Assume a transient error occurred, but that the account is on-chain and we can finish the creation process
                    try {
                        await wallet.saveAndMakeAccountActive(accountId);
                        await wallet.addLocalKeyAndFinishSetup(accountId, method.kind, recoveryKeyPair.publicKey);
                    } catch (e) {
                        showCustomAlert({
                            success: false,
                            messageCodeHeader: 'error',
                            errorMessage: e.message
                        });
                        redirectTo('/recover-account');
                    }
                }
```

Also some comments can be nice in code here re: 'what might this case have been caused by?' can be really useful for future investigations -- for example, since we know that the most likely case for !publicKeyExistsOnAccount is that someone else created the named account or someone had two tabs open and claimed it in a tab with a different seed phrase than the one they wrote down, it's worth noting that for future devs who are working in the code

Reply:
Thanks you ❤️